LibTest/io/Directory/renameSync_A04_t01.dart

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion Directory renameSync(String newPath)
+/// Synchronously renames this directory. Returns a Directory instance for the
+/// renamed directory.
+///
+/// If [newPath] identifies an existing directory, then the behavior is
+/// platform-specific. On all platforms, a [FileSystemException] is thrown
+/// if the existing directory is not empty. On POSIX systems, if [newPath]
+/// identifies an existing empty directory then that directory is deleted
+/// before this directory is renamed.
+///
+/// If newPath identifies an existing file or link the operation fails and a
+/// [FileSystemException] is thrown.
+///
+/// @description Checks that if [newPath] identifies an existing file, the
+/// operation fails and an exception is thrown.
+/// @author sgrekhov22@gmail.com
+
+import "dart:io";
+import "../../../Utils/expect.dart";
+import "../file_utils.dart";
+
+main() async {
+  await inSandbox(_main);
+}
+
+_main(Directory sandbox) async {
+  Directory srcDir = getTempDirectorySync(parent: sandbox);
+  Directory linkDir = getTempDirectorySync(parent: sandbox);
```

Comment:
Do you need to create a new directory for this? Can't you just create this in the parent?

Reply:
OK, I misunderstood what the test is doing, wrote a comment, and then deleted the comment. Here's a better one. I think.

We're creating a `srcDir` in the sandbox root, call it `src`, and then we're creating a `linkDir` in the sandbox root, call it `lDir`, and then we're creating a link `link` in the sandbox root such that `link` is a link to `lDir`. Then we're renaming `src` to `link`, and that's expected to incur a dynamic error.

The helper method `getTempLinkSync` will, AFAICS, create a symlink in the underlying OS file system, pointing to `target`. Then we rename. The `rename` documentation says that ..

> If `newPath` identifies an existing entity of the same type, that entity is removed first. If `newPath` identifies an existing entity of a different type, the operation fails and an exception is thrown.

So we'd expect the exception because we're renaming a `Directory` to a `Link`. It should not make any difference at all whether the link refers to a directory or to any other kind of file system entity, or even if it is dangling.

So it doesn't hurt to have `linkDir` and letting `link` be a symlink to `linkDir`, but it doesn't actually matter (e.g., `link` could be a link to `not_existing_file_system_entity`, and it should not give rise to any errors to create such a dangling link --- after all, we might create the target later).

Looks like that `linkDir` could indeed be omitted. ;-)

Reply:
Indeed. Fixed here and in the other places. Thank you!

Reply:
> If newPath identifies an existing entity of the same type, that entity is removed first. If newPath identifies an existing entity of a different type, the operation fails and an exception is thrown.

I think, this statement is removed. https://dart-review.googlesource.com/c/sdk/+/326862/4/sdk/lib/io/file_system_entity.dart

Reply:
@sgrekhov, very good, it sounds like there was something to double check here.

@brianquinlan, as I understand the situation at this point there was a discrepancy between the documented behavior and the actual behavior, and https://dart-review.googlesource.com/c/sdk/+/326862/4/sdk/lib/io/file_system_entity.dart#b279 was the CL where the documentation was corrected to match the actual behavior.

This means that `rename`/`renameSync` would never throw an exception because the `newPath` is a file system entity of a different kind, it would just overwrite the existing entity and succeed. (The operation could surely still throw an exception in the case where the running process doesn't have the required permissions, but that's a property which is shared by all the file system operations.)

But in that case I'd expect that the test should be adjusted such that it does not `Expect.throws` anything.

Reply:
@eernstg That's correct. The TL;DR is:
- `File`/`Link` `rename`/`renameSync` can replace other  `File`/`Link` but **not* `Directory`
- `Directory` `rename`/`renameSync` can **not** replace an existing `File`/`Link`/`Directory` **except** on POSIX where it can replace an empty `Directory`